Automatic certificates renewal

alexandre-allard opened the pull request:
**Component**: salt

**Context**:
In MetalK8s, we deploy a lot of certificates for interaction between components.
These certificates have an expiration date and must then be renewed, otherwise the services can no longer communicate.

**Summary**:
- Add a beacon configuration through role pillars for each deployed certificate to trigger an event when a certificate is expired or will expire.
- Add a reactor listening for the certificate expiration event, then based on information in role pillars, run specific sls to renew the expired certificates.

**Acceptance criteria**: 
Tests that every certificate is well renewed and that nothing breaks or every services that must be restarted are.

---

Closes: #1887

alexandre-allard committed:
salt: Use `certificates` for bootstrap role

This pillar entry will be consumed by the Salt
formulas configuring bootstrap role nodes,
the beacon and the reactor listening for
certificate expiration events.

If the path of an expired certificate matches one
in this list, the sls under `regen_sls` will be run.

Refs: #1887

alexandre-allard committed:
salt: Use `certificates` pillar for etcd role

This pillar entry will be consumed by the Salt
formulas configuring etcd, the beacon and the
reactor listening for certificate expiration
events.

If the path of an expired certificate matches one
in this list, the sls under `regen_sls` will be run.

Refs: #1887

alexandre-allard committed:
salt: Use `certificates` pillar for master role

This pillar entry will be consumed by Salt formulas
configuring master nodes, beacon and reactor
listening for certificate expiration events.

If the path of an expired certificate matches one
in this list, the sls under `regen_sls` will be run.

Refs: #1887

alexandre-allard committed:
salt, build: Add certs renewal orchestrate

This orchestrate will be called by the reactor
when it will receive an event for an expired
certificates.
It will run `sls` defined under `certs_renewal`
pillar entry for each expired certificate.

Refs: #1887

alexandre-allard committed:
salt, build: Add reactor for certs renewal

This reactor will be called when an expired
certificate event will be received.
It will then launch an orchestrate
`orchestrate.certs.renew`, passing the list
of expired certificates, to renew them.

Refs: #1887

alexandre-allard committed:
salt: Add certificate expiration reactor cfg

Set up the configuration in salt master cfg
for the certificate expiration reactor.

Refs: #1887

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1887-automatic-certs-renewal` into
`development/2.6`
* [w/2.7/feature/1887-automatic-certs-renewal](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/1887-automatic-certs-renewal) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #97436a4548dc93f3d8c96f380d0e82eea077abf6 on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #2c2bf0a422df67ef179760efeee2b43f801d5a49 on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1887-automatic-certs-renewal` into
`development/2.6`
* [w/2.7/feature/1887-automatic-certs-renewal](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/1887-automatic-certs-renewal) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

NicolasT commented:
This is not sufficient, since from what I can see it only checks 'plain file' certs. However, we have certs embedded in `KubeConfig`-style files, e.g. in `/etc/kubernetes/calico.conf` or `/etc/kubernetes/admin.conf` which should be checked and updated as well.

Also, I think there should be some form of documentation/notes somewhere explaining how this feature 'works' and how the various pieces work together.

alexandre-allard committed:
salt: Move kubeconfig validate method to a dedicated exec mod

This way this method can be called by anything to check
if the kubeconfig file needs to be regenerated.

The purpose is to be able to call it from a custom
beacon to check whether embedded certificates are
expired or not.

Refs: #1887

alexandre-allard committed:
salt: Add arg to pass remaining days to kubeconfig validate

It allows to dynamically pass the number of days
left to consider the certificates in kubeconfig
as expired.
This change is needed for the custom beacon used
to watch kubeconfig expiry as this parameter
will be configurable like it is done for other
beacons (e.g. cert_info).

Refs: #1887

alexandre-allard committed:
salt: Add new beacon module to check kubeconfig expiry

This beacon takes a list of kubeconfig as input
and checks whether they need to be renewed or not
triggering an event on Salt bus when needed.

To configure this beacon, a section must be
added, either in minion configuration or through
the pillar, as follows:

beacons:
  metalk8s_kubeconfig_info:
    - files:
        - /etc/kubernetes/calico.conf
        - /etc/kubernetes/admin.conf:
            notify_days: 30
    - interval: 86400
    - notify_days: 15

Default notify_days, if not provided, is 45.
It can be overridden for a specific kubeconfig as
shown above.

Refs: #1887

alexandre-allard committed:
salt: Add kubeconfig expiration reactor cfg

Set up the configuration in salt master cfg
for the kubeconfig expiration reactor.

Refs: #1887

gdemonet commented on salt/_beacons/metalk8s_kubeconfig_info.py:
```diff
@@ -0,0 +1,85 @@
+    if not isinstance(config, list):
+        return False, "{0}: it must be a list".format(error_msg)
+
+    _config = {}
+    list(_map(_config.update, config))
```
I guess this was copied from original Salt code :vomiting_face:. A cleaner, readable approach, would use a `for` loop:

```suggestion
    _config = {}
    for config_item in config:
        _config.update(config_item)
```

alexandre-allard replied:
Yes, it's what is done in each and every beacons, I wanted to do it the same way, but I can change it if you think it's too dirty. :)
I don't even understand why list of dict are used here, instead of a simple dict...

alexandre-allard replied:
Changed with a for loop :)

gdemonet commented on salt/_beacons/metalk8s_kubeconfig_info.py:
```diff
@@ -0,0 +1,85 @@
+    _config = {}
+    list(_map(_config.update, config))
+
+    if "files" not in config:
```
That's a typo I think:

```suggestion
    if "files" not in _config:
```

alexandre-allard replied:
It is, good catch.

alexandre-allard replied:
Fixed

gdemonet commented on salt/_beacons/metalk8s_kubeconfig_info.py:
```diff
@@ -0,0 +1,85 @@
+        return False, "{0}: 'files' key must be a list".format(error_msg)
+
+    mandatory_keys = {"name", "user"}
+    for kubeconfig in _config["files"]:
```
(minor) Could be useful to accumulate all errors from this list and print them at once.

alexandre-allard replied:
Done

gdemonet commented on salt/_beacons/metalk8s_kubeconfig_info.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def validate(config):
+    error_msg = "Configuration for {0} beacon is invalid".format(
+        __virtualname__
+    )
```
(minor) Could factorize some code with a helper function:

```python
BASE_ERROR_MSG = "Configuration for {0} beacon is invalid".format(
    __virtualname__
)
def _config_error(msg):
    return False, ": ".join([BASE_ERROR_MSG, msg])

def validate(config):
    if not isinstance(config, list):
        return _config_error("it must be a list")

    # ...
```

alexandre-allard replied:
Good idea, done.

gdemonet commented on salt/_beacons/metalk8s_kubeconfig_info.py:
```diff
@@ -0,0 +1,85 @@
+        if not __salt__['metalk8s_kubeconfig.validate'](
+                kubeconfig['name'], b64_ca_cert, apiserver,
+                kubeconfig['user'], remaining_days):
+            log.info("kubeconfig %s is out of date", kubeconfig['name'])
```
"out of date" is a bit strong, given the mechanism checks "in advance". Maybe something like:

```suggestion
            log.info("kubeconfig %s will expire in less than %d days",
                     kubeconfig['name'], remaining_days)
```

gdemonet replied:
However, the beacon actually checks more than the expiry date... And we don't get exactly why the file needs to be regenerated. Maybe something to improve in the future.

alexandre-allard replied:
Yes, maybe not talking about the expiration date is better.
Just tell that the kubeconfig needs to be regenerated.

gdemonet reviewed (commented):
Some comments on the kubeconfig-related parts.

bert-e commented:
# History mismatch



Merge commit #8e304e16a11c84b1990e5569562cc1d24b9f5f98 on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1887-automatic-certs-renewal` into
`development/2.6`
* [w/2.7/feature/1887-automatic-certs-renewal](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/1887-automatic-certs-renewal) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #b235d62cca1480711266e1a261955c8aca5dbab8 on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1887-automatic-certs-renewal` into
`development/2.6`
* [w/2.7/feature/1887-automatic-certs-renewal](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/1887-automatic-certs-renewal) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

alexandre-allard committed:
tests: Add unit tests for `metalk8s_kubeconfig` salt module

Refs: #1887

bert-e commented:
# History mismatch



Merge commit #319bee183ee710d91374cb880d386a4e97e00a62 on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
tests: Add .coverage to .gitignore

This file is generated by pytest when using
--cov* options for salt unit tests, so let's
ignore it to avoid polluting git command output
or committing it because of a careless mistake.

bert-e commented:
# History mismatch



Merge commit #edae2e9bf9edc5769f18324c64ec6467a3c3fb90 on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #9b4e2a389f61436bef793670f413961f3d042df9 on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
tests: Add test for certificates renewal beacon

This tests reconfigure the beacons and override
the pillar configuration to force the renewal
of all the certificates and kubeconfigs.

The goal is to ensure that beacons work well
and that nothing is broken in the cluster even
when everything is triggered at the very same
time.

Refs: #1887

alexandre-allard committed:
salt: Add defaults for `certificates`

These defaults will be merged with the pillar
and can be overriden, they'll be used by both
the certificate & kubeconfig expiry beacons,
the related reactor (certs renewal) and the
`x509.certificate_managed` state.

Refs: #1887

alexandre-allard committed:
salt: Add certificates expiry beacon setup sls

This sls setup the two beacons used to watch
certificates and kubeconfig expiry.
We also need to install pyOpenSSL package for
cert_info beacon to work.

Refs: #1887

alexandre-allard committed:
salt: Deploy certificates beacon on nodes with role `node`

Since this is needed on almost any node, let's
deploy the beacons on all nodes, if there is
no certificate to watch it will do nothing anyway.

Refs: #1887

alexandre-allard committed:
salt: Use `certificates` entry for calico

Replace hardcoded path for calico kubeconfig
in the related formulas, using the new
entries under certificates key in the
`defaults.yaml` file.

Refs: #1887

alexandre-allard committed:
salt: Add days_valid to kubeconfig state

This will allow to change the time a kubeconfig
is valid (by changing the embedded certificates
validity period).

Refs: #1887

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/1887-automatic-certs-renewal` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `feature/1887-automatic-certs-renewal`
* Rebase `feature/1887-automatic-certs-renewal` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

alexandre-allard commented on pillar/top.sls.in:
```diff
@@ -20,8 +20,19 @@ metalk8s-{{ version }}:
 
+  # Specific case for bootstrap when K8s node objects do not exist yet,
+  # we need to retrieve info from bootstrap role pillar.
+  not I@metalk8s:nodes:
```
Not really fond of this one, if someone has a better idea...
The purpose is to check if we're currently bootstrapping the node or not (and it must work if we launch the bootstrap again).
Maybe we could set up a grain before and then remove it after the bootstrap is done.

TeddyAndrieux replied:
I would prefer to set pillar value in bootstrap script so that we rely on the "classic" `for role in roles`.
Where does it break ? Because in the orchestrate we already set the pillar value so that we have the "bootstrap" role

gdemonet commented on salt/metalk8s/defaults.yaml:
```diff
@@ -113,3 +113,12 @@ coredns:
+  beacon:
+    notify_days: 45
+    interval: 86400  # once a day
+  days_valid: 365
+  days_remaining: 90
```
Should these really be set globally?

alexandre-allard replied:
You want to set these per certificate ? I also asked myself on this one, but I'm not sure it will be useful.
Otherwise we could do both and fall back on the global value if nothing is specified at a specific certificate level.
e.g. in `pillar/metalk8s/roles/bootstrap.sls`:
```
certificates:
  kubeconfig:
    salt-master:
      path: /etc/salt/master-kubeconfig.conf
      cn: salt-master-{{ grains.id }}
      days_remaining: 60
      days_valid: 180
      regen_sls:
        - metalk8s.salt.master.kubeconfig
[...]
```
Then in the formula `salt/metalk8s/salt/master/kubeconfig.sls` we could simply do:
```
[...]
    - days_valid: {{ certificates.kubeconfig['salt-master'].days_valid | default(certificates.days_valid) }}
    - days_remaining: {{ certificates.kubeconfig['salt-master'].days_remaining | default(certificates.days_remaining) }}
[...]
```
It does cost almost nothing to do, but I'm not really sure if it could be useful some day.

TeddyAndrieux replied:
Maybe just rename the root key (or subkey) so that here we only talk about "client certificates" and not CAs.
And for me, we may want at some point to change the one in some kubeconfig so it could be useful to override in the pîllar (or at least having different key for "client certificates", "kubeconfig certificates", "CA certificates")

gdemonet replied:
Yeah, maybe just build a macro to handle the defaulting, and indeed, maybe add a level to distinguish between cert "kinds" (CA - server - client - kubeconfig). Granted though, this will likely not be very useful to have this setting available per-cert, except for some of them (e.g. an admin could want to extend the duration of their `/etc/k/admin.conf` to avoid having to regenerate it every year).

alexandre-allard replied:
I've already changed this, we now have:
```
certificates:
  beacon:
    notify_days: 45
    interval: 86400
  client:
    days_remaining: 90
    days_valid: 365
    files:
      /path/to/client.crt:
        days_valid: 180
        [...]
  kubeconfig:
    days_remaining: 90
    days_valid: 365
    files:
      /path/to/kubeconfig.conf:
        [...]
```
It uses the configuration set under `certficates.[kubeconfig|client].files.$path` if any and fallback on `certificates.[kubeconfig|client]` otherwise.
I've put everything that's not CA nor kubeconfig under client because we handle them the same way, but maybe we want to distinguish client & server certificates ?

TeddyAndrieux replied:
Yes maybe add one section for server, right, it make sense

TeddyAndrieux commented on salt/metalk8s/beacon/certificates.sls:
```diff
@@ -0,0 +1,47 @@
+{%- if certificates.kubeconfig %}
+  {%- set files = [] %}
+  {%- for kubeconfig in certificates.kubeconfig.values() %}
+    {%- do files.append({
+      "name": kubeconfig['path'],
+      "user": kubeconfig["cn"],
+    }) %}
```
Why do we check CN only for kubeconfig ? 
To me, we should do the same for kubeconfig and "classic" certificates, and personally, I'm totally ok for not checking the CN in the beacons, the CN shouldn't change (except, maybe, during an upgrade and, in this case, it's handled by the upgrade)

alexandre-allard replied:
It's checked because it's the case in the function checking for the kubeconfig validity, so I followed what was done.
I wasn't sure if I should do it or not, to me it could be optional or we could retrieve it from the passed kubeconfig and so it would always match.

TeddyAndrieux replied:
To me, it should be optional and not checked by the beacon

alexandre-allard replied:
It's now optional and not checked by the beacon. :)

TeddyAndrieux commented on salt/metalk8s/kubernetes/cni/calico/configured.sls:
```diff
@@ -1,8 +1,12 @@
 {%- from "metalk8s/map.jinja" import networks with context %}
 
 {%- set kubernetes_service_ip = salt.metalk8s_network.get_kubernetes_service_ip() %}
+{%- set cn = salt['slsutil.renderer'](
+    string=certificates.kubeconfig.calico.cn
+) %}
```
This is really needed ? From what I remember pillar already get jinja rendered (but maybe it's only the case when you retrieve it from pillar directly and not in `map.jinja` :thinking: )

alexandre-allard replied:
You're right, just re tested. I don't know what I did but during a test it was not rendered, I may have failed something somewhere. :)
It even works when it comes from the `map.jinja`.

TeddyAndrieux commented on salt/metalk8s/salt/master/kubeconfig.sls:
```diff
@@ -1,18 +1,25 @@
 {%- from "metalk8s/map.jinja" import kube_api with context %}
 {%- from "metalk8s/map.jinja" import kubernetes with context %}
 
+{%- set cn = salt['slsutil.renderer'](
+    string=certificates.kubeconfig['salt-master'].cn
+) %}
```
ditto

alexandre-allard replied:
ditto :)

bert-e commented:
# History mismatch



Merge commit #614d25116b7889aa262c964ba8fa002ec763f69e on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
salt: Make check of the CN optional

We now only check the CN in the kubeconfig
certificate if the `expected_cn` is not None.
This is needed in order to avoid checking
the CN in the kubeconfig beacon which will
rely on this method.

Refs: #1887

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/1887-automatic-certs-renewal` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `feature/1887-automatic-certs-renewal`
* Rebase `feature/1887-automatic-certs-renewal` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# History mismatch



Merge commit #614d25116b7889aa262c964ba8fa002ec763f69e on the integration branch
`w/2.7/feature/1887-automatic-certs-renewal` is merging a branch which is neither the current
branch `feature/1887-automatic-certs-renewal` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/1887-automatic-certs-renewal` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.7/feature/1887-automatic-certs-renewal` with contents from `feature/1887-automatic-certs-renewal`
and `development/2.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.7/feature/1887-automatic-certs-renewal origin/development/2.7
 $ git merge origin/feature/1887-automatic-certs-renewal
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/feature/1887-automatic-certs-renewal
```

alexandre-allard committed:
salt: Use `certificates` entry for kubelet

Replace hardcoded path for kubelet kubeconfig
in the related formulas, using the new
entries under certificates key in the
`defaults.yaml` file.

Refs: #1887

alexandre-allard committed:
tests: Raise timeout in logging tests

Timeout has been raised because of a lot
of false positive in CI tests.
We now wait 40 * 5s (200s) for log to
show up in Loki from the logger Pod.

gdemonet commented on salt/_modules/metalk8s_kubeconfig.py:
```diff
@@ -16,7 +16,8 @@ def __virtual__():
              expected_api_server,
-             expected_cn):
+             expected_cn,
+             days_remaining=90):
```
Why change the value used by default (from 30 to 90)?

alexandre-allard replied:
To have the same default behavior as the `certificate_managed`, anyway we're not using this default value.

alexandre-allard committed:
Update CHANGELOG with certificate rotation feature

gdemonet commented on salt/metalk8s/addons/dex/certs/server.sls:
```diff
@@ -35,12 +36,18 @@ Create Dex server private key:
 Generate Dex server certificate:
   x509.certificate_managed:
-    - name: /etc/metalk8s/pki/dex/server.crt
+    - name: {{ certificates.server.files.dex.path }}
```
Since these files are configured in pillar now, we likely have other updates to do, like for this one:

- in `salt/metalk8s/addons/dex/deployed/tls-secret.sls`:

```diff
 data:
   tls.crt: "{{
-    salt['hashutil.base64_encodefile']('/etc/metalk8s/pki/dex/server.crt')
+    salt['hashutil.base64_encodefile'](certificates.server.files.dex.path)
     | replace('\n', '')
   }}"
   tls.key: "{{
```

alexandre-allard replied:
Oh I forgot some of those... too much certificates everywhere. :)

gdemonet commented on tests/test-certificates-beacon.sh:
```diff
@@ -0,0 +1,147 @@
+
+    crictl stop "$(get_salt_container)"
+
+    echo "Wait for Salt master to be ready..."
+    sleep 30
```
can we wait with `kubectl wait`? Could be more resilient

alexandre-allard replied:
Yes, totally.

gdemonet commented on CHANGELOG.md:
```diff
@@ -2,6 +2,12 @@
 
 ## Release 2.6.1 (development in progress)
 
+### Features Added
+- [#1887](https://github.com/scality/metalk8s/issues/1887) - All Kubernetes
+  kubeconfig, client and server certificates are now automatically regenerated
+  when close to the expiration date (less than 45 days)
+  (PR [#2914](https://github.com/scality/metalk8s/pull/2914))
```
:heart:

gdemonet reviewed (changes requested):
I'm afraid we need to make sure the cert/kubeconfig paths configured in pillar are also used in other places than where they are generated, in case we decide to change their location in the future.

NicolasT commented on pillar/metalk8s/roles/etcd.sls:
```diff
@@ -0,0 +1,17 @@
+          - metalk8s.kubernetes.etcd
+  server:
+    files:
+      etcd-peer:
```
Aren't peer certs both client and server?

NicolasT commented on pillar/metalk8s/roles/master.sls:
```diff
@@ -0,0 +1,38 @@
+        path: /etc/kubernetes/admin.conf
+        cn: kubernetes-admin
+        regen_sls:
+          - metalk8s.kubernetes.apiserver.kubeconfig
```
This is a slightly odd state name (granted, not introduced by this PR), since `admin.conf` has little to no relation to the apiserver being deployed somewhere.

alexandre-allard replied:
Yes, I agree, it seemed really weird to me when I saw that.
Maybe we should open an issue to tackle that once this PR is merged... or later.

NicolasT commented on pillar/top.sls.in:
```diff
@@ -25,3 +25,9 @@ metalk8s-{{ version }}:
+  # Apply on all nodes, except if the only role is CA.
+  {{ version_match }} and J@metalk8s:nodes:{{ grains.id }}:roles:^({{ roles | join("|") }})$:
+    - match: compound
+    - metalk8s.kubernetes.cni.calico
```
I'm not sure why all of a sudden we need these here?

NicolasT commented on pillar/metalk8s/kubernetes/kubelet.sls:
```diff
@@ -0,0 +1,8 @@
+    files:
+      kubelet:
+        path: /etc/kubernetes/kubelet.conf
+        cn: "system:node:{{ grains.id }}"
```
I'm surprised this wouldn't need some `o` as well? Reading on...

NicolasT replied:
So, reading on, it appears this is used to ensure the CN in the cert matches what's expected. However, I guess the same should then be done for `o`.

Related, not sure at this place we should actually validate `cn` and `o`: this is coded in the state that creates the cert in the first place. The beacon shouldn't really care and only ensure lifetime is acceptable. It has (and should not have) knowledge about who this cert is for, and call the applicable state(s) to (re)generate it.

alexandre-allard replied:
Oh yes there is also an `O` (but hard-coded in the formula), the `CN` here is a leftover (we needed in `metalk8s_kubeconfig.validate` function before we made it optional), but maybe we could put everything in this dict, so we have only one place with everything.

gdemonet replied:
Let's keep this "extracting all info about certs into pillar" for a later time. Would make sense to remove the CN for now, since it's not used, and see later how we want to improve this.

alexandre-allard replied:
I've removed this from the pillar.

NicolasT commented on pillar/metalk8s/kubernetes/cni/calico.sls:
```diff
@@ -0,0 +1,8 @@
+    files:
+      calico:
+        path: /etc/kubernetes/calico.conf
+        cn: "{{ grains.id }}"
```
I think we should use a more descriptive CN for this one. Or would that impact role bindings?

alexandre-allard replied:
Don't know, I can give it a try, but to me, it's not directly related to this PR.

gdemonet replied:
Let's have a ticket to track these "cert improvements" / debt.

NicolasT commented on pillar/metalk8s/roles/bootstrap.sls:
```diff
@@ -0,0 +1,32 @@
+  kubeconfig:
+    files:
+      salt-master:
+        path: /etc/salt/master-kubeconfig.conf
```
It's, in a way, kinda odd to have this one in `/etc/salt` and e.g., the `etcd` client cert in `/etc/kubernetes/pki/etcd`.

gdemonet replied:
Right, both should have gone in `/etc/salt/pki`, my bad. Will open a debt ticket.

NicolasT commented on pillar/metalk8s/roles/master.sls:
```diff
@@ -0,0 +1,38 @@
+      apiserver:
+        path: /etc/kubernetes/pki/apiserver.crt
+        regen_sls:
+          - metalk8s.kubernetes.apiserver
```
I believe this state (and many others we use for this purpose) do quite a bit more than setting up a new cert and restarting a service if applicable. Is it acceptable this is triggered by a beacon? CC @gdemonet

alexandre-allard replied:
Yes, should rather be `metalk8s.kubernetes.apiserver.installed` at least.

gdemonet replied:
Strictly speaking, the certificate generation happens in `.certs.server`, but all `.certs` are `include`d in `.installed`. And I think we need `.installed` to make sure the new cert triggers an update of the k-a static Pod.

NicolasT commented on salt/_modules/metalk8s_kubeconfig.py:
```diff
@@ -0,0 +1,106 @@
+        return False
+
+    # Verify that the mode is 600
+    if stat.S_IMODE(os.stat(filename).st_mode) != 0o600:
```
Strictly speaking `0400` would be just fine as well, so do we need to fix a specific value here?

NicolasT replied:
Rationale: if we make them `0400` in the state, then the beacon would always trigger, right?

alexandre-allard replied:
Yes, that's it or we could also change the mode used in `metalk8s_kubeconfig.managed` to set mode `0400`, so it would only change the mode and renew certificates for all kubeconfigs during an upgrade when coming from a version without this change.
Or we could, slightly change how this state behave and instead of returning directly when the `validate` succeed, we could return a `file.managed` only enforcing the mode to be `0400` and remove the file permission check from `metalk8s_kubeconfing.validate`.
We could also change this state to let Salt handle the changes in `file.serialize` (just retrieving certificate/key from kubeconfig instead of creating them when they're up to date).

alexandre-allard replied:
Opened an issue to track this #2958

NicolasT commented on salt/metalk8s/orchestrate/certs/renew.sls:
```diff
@@ -0,0 +1,18 @@
+{%- set target_pillar = salt.saltutil.runner(
```
Why exactly does this need to be an orchestration? :thinking: Is this required for a reactor to work?

alexandre-allard replied:
Not exactly required, but in the good practices, they say:

> Matching and rendering Reactor SLS files is done sequentially in a single process. For that reason, reactor SLS files should contain few individual reactions (one, if at all possible). Also, keep in mind that reactions are fired asynchronously (with the exception of caller) and do not support requisites.

source: https://docs.saltstack.com/en/latest/topics/reactor/#best-practices-for-writing-reactor-sls-files

I assumed an orchestrate would be better.

NicolasT commented on salt/metalk8s/salt/master/files/master-99-metalk8s.conf.j2:
```diff
@@ -59,3 +59,9 @@ kubernetes.kubeconfig: {{ kubeconfig }}
 {%- endif %}
+
+reactor:
+  - 'salt/beacon/*/watch_certificates_expiry/':
```
Should we namespace our event tags? CC @gdemonet

gdemonet replied:
We can't namespace beacon events, I believe. IIUC, the format will always be:

```
salt/beacon/$minion_id/$beacon_config_name/$tag
```

gdemonet replied:
If we really want to, we can add a reactor to transform event tags, but feels unnecessary.

alexandre-allard replied:
We could kind of namespace our events, just pre-prending beacon config name with metalk8s:
```
beacons:
  metalk8s_watch_certificates_expiry:
  - files:
    - /etc/metalk8s/pki/dex/server.crt
    [...]
  - notify_days: 45
  - disable_during_state_run: true
  - interval: 86400
  - beacon_module: cert_info
```

gdemonet replied:
Oh yeah, I meant they'll always live under `salt/beacon/$minion_id`. Using prefixed names can be useful if we expect other beacons to be configured, good idea.

NicolasT commented on scripts/upgrade.sh.in:
```diff
@@ -75,6 +75,8 @@ upgrade_bootstrap () {
     repo_endpoint="$($SALT_CALL pillar.get \
         metalk8s:endpoints:repositories --out txt | cut -d' ' -f2- )"
+    certificates=$($SALT_CALL pillar.get \
+        certificates --out txt | cut -d' ' -f2-)
```
We namespace all other pillar values. Shouldn't this as well?

alexandre-allard replied:
We're not namespacing everything (e.g. networks, proxies, signing policies), still we could.

gdemonet replied:
If we want to, warrants another ticket.

alexandre-allard committed:
salt: Add `certificates` entry for `node` role

This pillar entry will be consumed by the Salt
beacon configuration formula.
This beacon watches certificate expirations.

Refs: #1887

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.7/feature/1887-automatic-certs-renewal` with contents from `feature/1887-automatic-certs-renewal`
and `development/2.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.7/feature/1887-automatic-certs-renewal origin/development/2.7
 $ git merge origin/feature/1887-automatic-certs-renewal
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/feature/1887-automatic-certs-renewal
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1887-automatic-certs-renewal` into
`development/2.6`
* [w/2.7/feature/1887-automatic-certs-renewal](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/1887-automatic-certs-renewal) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet

alexandre-allard commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
